### ISSUE TITLE
Improve Text Drawing Efficiency

### ISSUE THREAD

**Issue by philplckthun** [rogual/neovim-dot-app#105]
I really like this project :) I'd love to have a clean and sleek container app for Neovim, but I'm having some performance issues. Quick dive in the code showed that the most inefficient part is the glyph / text rendering. The app often has to update 150 glyphs fairly often (this can surely be minimised. I doubt that that many even change that often).

The clipping rect takes about 0.1ms which sums up in the case of 150 glyphs to a redraw time of 15ms. This is way too inefficient. I could get that time down to a lot less by just not clipping, but as the code comment says this breaks the font rendering and makes it look glitchy. (On a side note: The left-to-right unicode force also slows down the redraw time a bit)

I'm not at all experienced at Cocoa, but when I look at some of Apple's docs and at MacVim's implementation it seems that the way to go is to use a text storage and a layout manager. This is of course much more complex than just drawing some glyphs in rects, but I think this could make a huge difference.

TLDR: redraw.mm needs to be refactored

**Comment by gglanzani** [comment]
+1

ATM I'm using neovim in iTerm2 + tmux (with full color and the nice cursor) as neovim-dot-app is fast when starting it, but then slowly start crawling.

**Comment by rogual** [comment]
Good insights, thanks. The slow drawing is our #1 issue right now as far as I'm concerned. Hopefully we can get it fixed. Unfortunately I'm not a Cocoa expert either.

Doing things the MacVim way may turn out to be the right approach, but MacVim doesn't render text quite how I want it — for instance, it doesn't use fallback fonts to render foreign scripts, so that alignment test block in the comment in redraw.mm renders with lots of missing chars in MacVim.

Ideally I'd like the font rendering to be both correct *and* fast, if at all possible.

**Comment by philplckthun** [comment]
@rogual I think if we want to preserve the current font rendering, then we somehow need to determine the actual line height, so we can get an actual height of the glyph tiles. Then I guess if we want to keep it efficient we don't want to rerender those tiles every time they change. And we might want to filter out redrawing glyphs that didn't actually change. And we may not need a layout manager, but we have to cache the glyphs in an array or something similar.

**Comment by philplckthun** [comment]
@rogual This is very weird: https://github.com/philplckthun/neovim-dot-app/commit/9f9d5f6

I was basically able to get the redraw time down dramatically, by removing redundant code. By removing the redundant code I was able to remove the clipping, which was the most expensive operation.

Wutafuq?! Why does this work?!

**Comment by rogual** [comment]
Unfortunately, making that change appears to bring back bug https://github.com/rogual/neovim-dot-app/issues/4 for me.

**Comment by philplckthun** [comment]
@rogual As I see it I reverted some changes of #58 by chance. My best guess is that that's not the correct way to fix the line height bug, but I'm not sure where to search for a better fix.

**Comment by rogual** [comment]
Yeah, there's almost certainly a better way. I'll probably look into Core Text next.

**Comment by philplckthun** [comment]
@rogual I guess we can still fix it as it is. I'm searching for some bugs in the origin calculations, already found one. I believe it's possible to get this working correctly.

**Comment by philplckthun** [comment]
@rogual Pushed some commits that fix #58 again. So this is a working performance improvement. However there still needs to be an option to set a line padding / height. But overall I think I can close this issue by sending a PR of my changes.

**Comment by bambu** [comment]
@philplckthun 
This definitely a performance improvement. However, it seems to have some bugs. 

When loading a new file (in this  case src/redraw.mm) some unicode doesn't show up properly. This is easily seen in vim-airline.  Secondly, scrolling down (and not in the opposite direction) causes it to have background lines and be blurry. The background line seems to be from airline (changed the colorscheme and the lines still matched the airline color).

See the following picture for example

<img width="601" alt="screen shot 2015-08-27 at 6 32 33 pm" src="https://cloud.githubusercontent.com/assets/564972/9534922/6da142f4-4cea-11e5-8467-21b27400cc52.png">

Both issues are fixed after the window loses focus.

**Comment by philplckthun** [comment]
@bambu I'll probably still need to limit some drawing areas. And fix up the maths for the scrolling. But I'm thinking about rewriting the app's codebase in Swift and starting over with CoreText. What do you think about that, @rogual?

**Comment by philplckthun** [comment]
@bambu @rogual Ok, this bug was pretty dumb. I forgot to readd a ceil to the font height calculation. Can't have enough of those I guess...

This should be good to go

**Comment by bambu** [comment]
Nicely done @philplckthun 

Latest commit works a lot better. Passed all of the visual tests that I ran on it previously. I would suggest removing the "Took too long" messages.

**Comment by philplckthun** [comment]
@bambu Forgot about those. Added another commit

**Comment by rogual** [comment]
> I'm thinking about rewriting the app's codebase in Swift and starting over with CoreText. What do you think about that, @rogual?

Good luck!

**Comment by philplckthun** [comment]
@rogual Can be closed for now?

**Comment by rogual** [comment]
Yeah.
